### PR TITLE
Add torch.promote_types function

### PR DESCRIPTION
**Pull request by nairbv**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26051 Add type promotion support in sparse tensors
* **#26655 Add torch.promote_types function**

Differential Revision: [D17556196](https://our.internmc.facebook.com/intern/diff/D17556196)

### CODE REVIEW

File: torch/_torch_docs.py [hunk 1]
```diff
@@ -3976,6 +3976,26 @@ def merge_dicts(*dicts):
     tensor([-0.2018, -0.2962, -0.0821, -1.1831])
 """.format(**single_dim_common))
 
+add_docstr(torch.promote_types,
+           r"""
+promote_types(type1, type2) -> dtype
+
+Returns the :class:`torch.dtype` with the smallest size and scalar kind to which
+`type1` and `type2` may be safely cast. See type promotion :ref:`documentation <type-promotion-doc>`
```
Comment (gchanan) [inline torch/_torch_docs.py]: "safely cast" isn't really defined here and I would argue are casting rules aren't really safe (i.e. int32 + float32) -> float32 can lose precision.  Can you reword this somehow?

Comment (nairbv) [reply torch/_torch_docs.py]: updated to a more specific explanation.

File: torch/_torch_docs.py [hunk 2]
```diff
@@ -3976,6 +3976,27 @@ def merge_dicts(*dicts):
     tensor([-0.2018, -0.2962, -0.0821, -1.1831])
 """.format(**single_dim_common))
 
+add_docstr(torch.promote_types,
+           r"""
+promote_types(type1, type2) -> dtype
+
+Returns the :class:`torch.dtype` with the smallest size and scalar kind that is
+not smaller or of lower kind than `type1` or `type2`. See type promotion
```
Comment (gchanan) [inline torch/_torch_docs.py]: nit: not smaller nor of a lower kind than either

**Review by gchanan — APPROVED** [review]

**Comment by gchanan** [comment]
does this work with JIT?

**Comment by nairbv** [comment]
> does this work with JIT?

Seems like it works as much as scalar types in general work (they're treated as ints):
```
>>> code='''
... def foo():
...     return torch.promote_types(torch.int, torch.double)
... def bar():
...     return torch.double
... '''
>>> cu=torch.jit.CompilationUnit(code)
>>> cu.foo()
7
>>> cu.bar()
7
```

**Pull request closed without merging**